fvtest/tril/tril/parser.cpp

```diff
@@ -0,0 +1,560 @@
+    return (tokenIt->getType() == Token::DOUBLE_ || tokenIt->getType() == Token::INT ||
+            tokenIt->getType() == Token::STRING_ || tokenIt->getType() == Token::HEXINT ||
+            tokenIt->getType() == Token::ZERO || tokenIt->getType() == Token::ID);
+}
+
+/* Build the value of node based on the given input
+ * 
+ * @param token The token to be build as node value
+ * @return The single node value
+ */
+ASTValue* buildNodeValue(Token token) {
+    Token::Type tokenType = token.getType();
+    std::string tokenValue = token.getValue();
+    if (tokenValue.size() > 1 && tokenType == Token::STRING_) { 
+        tokenValue = tokenValue.substr(1, tokenValue.size() - 2);
+    }
+    char * cstr = new char[tokenValue.size() + 1];
+    strcpy(cstr, tokenValue.c_str());
+    if (tokenType == Token::STRING_ || tokenType == Token::ID) {
+        return createStrValue(cstr);
+    } else if (tokenType == Token::DOUBLE_) {
+        double tmpdouble = std::atof(cstr);
+        return createFloatingPointValue(tmpdouble);
+    } else if (tokenType == Token::INT || tokenType == Token::ZERO) {
+        return createIntegerValue(std::strtoull(cstr, NULL, 10));
+    } else if (tokenType == Token::HEXINT) {
+        return createIntegerValue(std::strtoull(cstr, NULL, 16));
+    } else {
+        throw ParserFailure("Do not have token type with value " + tokenValue + " as value type");
+    }
+}
+
+/* Parse and build the list of the node value
+ * 
+ * @param tokenIt The tokens contains the node value info
+ * @return The node value linked list
+ */
+ASTValue* parseValueList(TokenIter &tokenIt) {
+    ASTValue* currentValue = NULL;
+    while (tokenIt->getType() != Token::RPAREN_SQ) {
+        try {
+            if (currentValue == NULL && tokenIt->getType() != Token::COMMA) {
+                currentValue = buildNodeValue(*tokenIt);
+            } else if (currentValue && tokenIt->getType() != Token::COMMA) {
+                appendSiblingValue(currentValue, buildNodeValue(*tokenIt));
+            }
+        } catch (ParserFailure &f) {
+            std::cerr << f.what() << std::endl;
+        }
+        ++tokenIt; // consume to be COMMA, the node value, or the RPAREN_SQ
+    }
+    return currentValue;
+}
+
+/* Parse and build the list of the node argument
+ * 
+ * @param tokenIt The tokens contains the node argument info
+ * @return The node argument linked list
+ */
+ASTNodeArg* parseArgList(TokenIter &tokenIt) {
+    ASTNodeArg* currentArg = NULL;
+    ASTNodeArg* argList = NULL;
+
+    while (tokenIt->getType() != Token::LPAREN && tokenIt->getType() != Token::RPAREN) {
+        if (tokenIt->getType() == Token::ID && tokenIt.peek().getType() == Token::EQUALS) {
+            char * node_args_name = new char[tokenIt->getValue().size() + 1];
+            strcpy(node_args_name, tokenIt->getValue().c_str());
+            ++tokenIt; // consume EQUALS
+            ++tokenIt; // consume to be LPAREN_SQ(the start of value list) or single node value
```

Comment:
Are we going to write the token type before or after adding `++tokens`? Yes, the tokens before adding are ID and EQUALS (sorry, I thought we consume the type after adding them, since if conditions already limit the current token type, so we have to consume the next one)

Reply:
Oh... just noticed your comment to this topic, will change them into current token type :)